app/src/org/commcare/dalvik/geo/HereFunctionHandler.java

```diff
@@ -0,0 +1,149 @@
+package org.commcare.dalvik.geo;
+
+import android.Manifest;
+import android.content.Context;
+import android.content.pm.PackageManager;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.os.Looper;
+import android.support.v4.content.ContextCompat;
+import android.util.Log;
+
+import org.commcare.dalvik.activities.EntitySelectActivity;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.model.condition.EvaluationContext;
+import org.javarosa.core.model.condition.IFunctionHandler;
+import org.javarosa.core.model.data.GeoPointData;
+import org.odk.collect.android.utilities.GeoUtils;
+
+import java.util.Set;
+import java.util.Vector;
+
+/**
+ * HereFunctionHandler is a class that can be used to evaluate the here() XPath expression.
+ * In any context where here() needs to be evaluated, create a HereFunctionHandler and
+ * call addFunctionHandler to add it to the XPath evaluator.
+ *
+ * In addition, an EntitySelectActivity can register itself to be refreshed whenever
+ * a new value of here() is obtained (whenever the location changes).
+ *
+ * @author Forest Tong
+ */
+public class HereFunctionHandler implements IFunctionHandler, LocationListener {
+    public static final String HERE_NAME = "here";
+    private GeoPointData location;
+
+    private LocationManager mLocationManager;
+    private Set<String> mProviders;
+
+    // This flag keeps track of whether the location manager needs to be set up and subscribed to.
+    private boolean locationUpdatesRequested = false;
+
+    private Context context;
+    // If there are more general uses for HereFunctionHandler, the type of this field can be
+    // generalized to a listener interface.
+    private EntitySelectActivity entitySelectActivity;
+
+    public HereFunctionHandler() {
+        this.context = CommCareApplication._().getApplicationContext();
+    }
+
+    public void registerEntitySelectActivity(EntitySelectActivity entitySelectActivity) {
+        this.entitySelectActivity = entitySelectActivity;
+    }
+
+    // Setting up and subscribing to the location manager is delayed until here() is actually
+    // evaluated.
+    public Object eval(Object[] args, EvaluationContext ec) {
+        if(!locationUpdatesRequested) requestLocationUpdates();
+        if(location == null) return "";
```

Comment:
Yeah, might be good to setup our codestyle config in Android Studio, which will help you do this automatically (ctrl-alt-l in Android Studio). See https://confluence.dimagi.com/display/MD/Code+Conventions#CodeConventions-AndroidStudioCodingConventionSettings. I can help you out with it, because it is a little finicky.